tests/lax_test.py

```diff
@@ -901,28 +903,15 @@ def testSqueezeShapeCheck(self, inshape, dimensions, error_type, err_msg):
     with self.assertRaisesRegex(error_type, err_msg):
       lax.squeeze(x, dimensions=dimensions)
 
-  @parameterized.named_parameters(jtu.cases_from_list(
-      {"testcase_name": "_inshape={}_dimensions={}".format(
-          jtu.format_shape_dtype_string(arg_shape, onp.float32), dimensions),
-       "arg_shape": arg_shape, "dimensions": dimensions,
-       "rng_factory": rng_factory}
-      for arg_shape, dimensions in [
-          [(1,), (0,)],
-          [(1,), (-1,)],
-          [(2, 1, 4), (1,)],
-          [(2, 1, 3, 1), (1,)],
-          [(2, 1, 3, 1), (1, 3)],
-          [(2, 1, 3, 1), (3,)],
-      ]
-      for rng_factory in [jtu.rand_default]))
-  def testSqueeze(self, arg_shape, dimensions, rng_factory):
-    rng = rng_factory(self.rng())
-    args_maker = lambda: [rng(arg_shape, onp.float32)]
-    op = lambda x: lax.squeeze(x, dimensions)
-    numpy_op = lambda x: lax_reference.squeeze(x, dimensions)
-    self._CompileAndCheck(op, args_maker)
-    self._CheckAgainstNumpy(op, numpy_op, args_maker)
-    check_grads(op, args_maker(), 2, ["fwd", "rev"], eps=1.)
+
+  @primitive_harness.parameterized(primitive_harness.lax_squeeze)
+  def testSqueeze(self, harness: primitive_harness.Harness = primitive_harness.lax_squeeze[0]):
+    self._CompileAndCheck(harness.dyn_fun,
```

Comment:
This change seems very heavyweight for what it achieves and in my mind does significant harm to the readability of the tests. In particular, the test harness is no longer next to the test, and I would not want to maintain the result of doing this widely.

Couldn't we achieve the same end goal (test both JAX and JAX-to-TF implementations of each primitive) by extending `_CompileAndCheck` (or adding a sibling to `_CompileAndCheck` that is also called from the lax tests?) The goal of `_CompileAndCheck` is essentially to compare two different implementations of a function (jitted and unjitted). Wouldn't it be simpler to extend it to compare three?

Another option might be to parameterize the test class as a whole (e.g., subclass it for each backend or something like that.)

Reply:
We've also discussed that as part of the omnistaging change it may be possible to switch out which implementations of primitives you get under a scope (i.e., what the `impl` rule does). That may be helpful here, because then you can achieve much the same end by running exactly the same code in a scope (e.g., via a different test subclass that wraps test invocations or something like that.)

Reply:
I agree that it is not great that this moves the harness away from the test code. But if you want to use the harness in more than one test, in at least one of them it is going to be away from test code. First I tried to keep lax_test as is (with harness + check together), and inherit the harness from it in jax_to_tf tests. This is not easy because lax_test.py is outside the jax package source tree, and not easily importable. Hence, the idea of primitive_harness inside the jax source tree.  

I also tried to inherit the whole LaxTest class and then try to reuse it by hijacking various pieces of it. One of the attempts was to actually hijack the impl rule. That is quite messy, and couples the tests in complicated ways.

What I want to achieve is to separate the harness (the set of inputs covering each primitive), from what you do with it.

Reply:
My gut agrees with @hawkinsp that this will be hard to maintain.

The reason why I find this hard to read is that it separates the definition of the "harness" from the test itself. Now there are two separate files to keep in sync for each LAX function. This seems like it could be quite painful to maintain/upgrade.

I would be curious to see a sketch of what the other approach with inheritance looks like. Maybe we'll be convinced after we see how messy the alternative is :)

Reply:
This is a good idea, I can try to prepare another PR with an alternative. 

However, I do think that in this case there is a separation of concerns between the harness and the check. The harness defines a set of inputs to the primitive that we believe covers its semantics. This is separate from what you do with it: you compare compiled and uncompiled code, you compare with numpy, etc. Because of this separation I think it is not too bad to have them in separate places. 

Consider this code in lax_test.py
```
  @primitive_harness.parameterized(primitive_harness.lax_pad)
  def testPadAgainstNumpy(self,  harness: primitive_harness.Harness):
    self._CheckAgainstNumpy(lax_reference.pad,
                            harness.fun,
                            lambda: harness.args_maker(self.rng()))
```

I can understand this test without needing to know what sets of inputs are in `primitive_harness.lax_pad`. Similarly, I can read the `primitive_harness.lax_pad` without needing to know what checking I am going to do with it.

Reply:
However in many of these tests (perhaps less so in `lax_test.py`) the contract is not a direct match to a primitive's API, but a contract with the test function itself. So separating the two harms readability.

Reply:
I agree that this is separately understandable, but there are also a lot of repeated references to the specific function, which suggests that it someone specialized to the inputs.

Is there a reason for writing separate test methods for each function? It seems like it might be possible to consolidate these, e.g., into something like the following:
```
  @primitive_harness.parameterized(
      primitive_harness.lax_pad,
      primitive_harness.lax_squeeze,
      ...)
  def test(self, harness: primitive_harness.Harness):
    self._CheckAgainstNumpy(harness.reference_fun
                            harness.fun,
                            lambda: harness.args_maker(self.rng()))
    if harness.differentiable:
      jtu.check_grads(...)
```

Here's it's clear that nothing is particular to `pad` and all that needs to be kept in sync is the list of harnesses to check.

Reply:
To @shoyer: Absolutely we can consolidate the CheckAgainstNumpy tests for multiple harnesses. Otherwise, I suspect that there will be a lot of identical test functions with a different harness each. I did not do that yet because I wanted to just try out this idea and discuss it before I put too much effort in it.

Reply:
Re @hawkinsp: You are probably right, I do not intend to hoist all harnesses out. Only those that are about covering a primitive's semantics.

Reply:
Meanwhile I have discovered that lax.pad does not have test coverage for negative padding, and in fact lax_reference.pad is not correct for that case. I will definitely add a test case for that in jax_to_tf, but would be nice to have a shared harness.

Reply:
Implemented this idea separately for jax_to_tf tests (#3376). We can extend it later to lax_test.py as in this PR.